scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -622,11 +668,12 @@ object Execution {
                   Future.successful(Map.empty)
               }
 
-            failFastZip(otherResult, localFlowDefCountersFuture).map {
-              case (lCounters, fdCounters) =>
-                val summedCounters = (fdCounters :: lCounters).reduce(_ ++ _)
-                (fn(conf, mode), summedCounters)
-            }
+            val bothFutures = failFastZip(otherResult, localFlowDefCountersFuture)
```

Comment:
First note, I didn't change that, I just captured the future into a name so I could use a for comprehension and have it read nicer, in my opinion.

About how this works, I hope the code is somewhat readable, but it is not simple. This is where a lot of the hard work of Execution is happening and it got more complicated over time due to finding and fixing issues with its use at Twitter. The thing we are working very hard to do here is not ever re-run a ToWrite aside from also not re-running ANY execution. The problem is, if things are zipped differently, the same ToWrite might end up in two different Executions. So we look through for all the ToWrites we didn't yet run, we run those in one call to the Writer, and we merge the counter results from the other previous writes (otherwise the counter results are wrong).